README
============================
Enrich-0.2 documentation
============================
:Authors: Douglas M. Fowler, Carlos L. Araya, Wayne Gerards
:Homepage: http://depts.washington.edu/sfields/software/enrich/
:Organization: University of Washington
:Address: Department of Genome Sciences, Box 355065, Seattle, WA, 98195
:Copyright: Douglas M. Fowler and Carlos L. Araya, 2010-2011
:License: FreeBSD
:Contact: Douglas M. Fowler (dfowler at uw dot edu)

============
Installation
============
Enrich retains core functionality when run without any :ref:`dependencies <dependencies>` in a pure Python environment. Thus, Enrich can be run by executing the ``enrich.py`` script located in the Enrich tarball without any taking any other steps. 

For convenience, Enrich is can be installed using `easy_install <http://peak.telecommunity.com/DevCenter/EasyInstall>`_ which is part of `setuptools <http://peak.telecommunity.com/DevCenter/setuptools>`_. Please obtain the current version of Enrich `here <http://depts.washington.edu/sfields/software/enrich/>`_. After downloading Enrich, install it by typing::

	easy_install Enrich-|version|.tar.gz
	
If you do not have setuptools, running the ``ez_setup.py`` script in the source tarball will automatically download and install setuptools followed by Enrich.

Installation on a cluster
-------------------------
If you are using a cluster and do not have root access, ``easy_install`` may fail because it does not have write access to the default installation directories.  You can overcome this limitation by using the ``-d`` easy_install option, which allows you to specify an installation directory to which you do have access.  The install directory will need to be in your *$PYTHONPATH*, and you will also need to add it to your *$PATH* if you wish to use the ``enrich`` executable, which will be installed in the specified directory.

.. _dependencies:

Dependencies
------------
Enrich will identify and enumerate variants as well as calculate enrichment ratios without any dependencies. However, Enrich depends upon several common Python packages to perform optional tasks. Many of these dependencies are available as precompiled binaries, and it is recommended that these precompiled binaries be used whenever possible.  Below, each the purpose of each dependency is explained along with recommended versions and brief guidance for installation. Installing the dependencies in the order listed is recommended.

DRMAA
^^^^^
:What it's used for: DRMAA is required to interface with OGE
:What versions have been tested: DRMAA-0.4b3
:How is it installed?: DRMAA can be installed using easy_install by typing ``easy_install drmaa==0.4b3`` 
    
More information is available `here <http://code.google.com/p/drmaa-python/>`_

Numpy and Scipy
^^^^^^^^^^^^^^^
:What it's used for: Numpy and Scipy are used for generating visualizations and for calculating q-values for enrichment and depletion
:What versions have been tested: numpy-1.4.1, scipy-0.8.0 and 0.9.0
:How is it installed?: Precompiled binaries are available for Numpy and Scipy for OS X, Windows and some flavors of Unix. These, as well as source code and instructions for installation are available `here <http://new.scipy.org/download.html>`_.


Biopython
^^^^^^^^^ 
:What it's used for: Biopython is used for performing the optional Needleman-Wunsch alignment of read pairs
:What versions have been tested: biopython-1.5.2 and 1.5.5
:How is it installed?: Biopython can be installed using easy_install by typing ``easy_install biopython==1.55b``

More information is available `here <http://biopython.org/DIST/docs/install/Installation.html>`_

Matplotlib
^^^^^^^^^^
:What it's used for: matplotlib is used to generate visualizations
:What versions have been tested: matplotlib-1.0.1
:How is it installed?: Precompiled binaries are available for matplotlib for OS X and Windows and are available `here <http://sourceforge.net/projects/matplotlib/files/matplotlib/matplotlib-1.0.1/>`_.  Detailed instructions for installation, including from source code, are available `here <http://matplotlib.sourceforge.net/users/installing.html>`_.


CHANGELOG
2011-09-15  Douglas M. Fowler   <dfowler@uw.edu>
    *v0.2 includes a nubmer of major updates including:
        -removal of R as a dependency
        -implementation of visualizations using matplotlib
        -implementation of Scipy/Numpy to perform Poisson tests
        -expurgation of the usage of most Biopython functions (none are now required)
        -significant changes in implementation to reduce runtime to ~2h for a typical data set
        
    *documentation updated to reflect the much simplified installation process

enrich/map_unlink.py
#!/usr/bin/env python
'''
map_unlink: the map_unlink module uses unique variant counts generated by map_counts to calculate unlinked mutation frequencies for all possible single mutants from all variants. 

Variants with multiple mutations are unlinked; the variant frequency is added to each of the single mutants comprising the multiply mutated variant.  Map_unlink can be invoked in multipe modes.  Currently supported modes are counts and wild_counts. The difference between counts and wild_counts is that, for counts only positions that differ from wild type (i.e. mutations) are ennumerated whereas for wild_counts, all positions for each variant are included in the calculation regardless of the identity of the amino acid. In other words, in counts mode the map_unlink calculation at each position yields the frequency of mutants amino acids only whereas the wild_counts mode yields frequency information for all amino acids. Note that this module contains many currently undocumented modes other than wild_counts and counts. They are fully functional but have not been extensively tested.
'''

import sys, os, time #import standard libraries

__author__ = "Douglas M. Fowler"
__copyright__ = "Copyright 2011"
__credits__ = ["Douglas M Fowler", "Carlos L. Araya"]
__license__ = "FreeBSD"
__version__ = "0.2"
__maintainer__ = "Douglas M. Fowler"
__email__ = "dfowler@uw.edu"

def main(path, infile, molecule, mode, size, grid = 'L'):

    if grid != 'L':
        print time.asctime(time.localtime())
        print path, infile, molecule, mode, size
    
    try:
        #set input path and output file:
        f_infile = open(path + 'data/output/' + infile, 'U')
        f_outfile = open(path + 'data/output/' + 'unlink' + '_' + mode + '_' + infile.split('_', 1)[1], 'w')
    
    except:
        print 'Error: could not open input file'
        return 1
    
    try:
        size = int(size)
            
    except:
        print 'Error: integer-only parameters were not integers'
        return 1
    
    #set characters for molecule type:
    symbols = []
    if molecule == 'PRO':
        symbols = ['A','C','D','E','F','G','H','I','K','L','M','N','P','Q','R','S','T','V','W','Y','*']
        
    elif molecule == 'DNA':
        symbols = ['A','T','G','C']
    
    else:
        print 'Error: molecule type invalid'
        return 1    
    #print the header:
    print >> f_outfile, '\t'.join(['position'] + symbols)
    
    #generate keys for internal dictionary:
    l = f_infile.readline()
    data = {}
    for i in range(0,size):
        data[i] = {}
        for symbol in symbols:
            data[i][symbol] = []
    
    while l:
        l = f_infile.readline()
        line = l.rstrip().split('\t')
            
        if len(line) > 4 and mode == "wild_counts":
            sequence, y, empty = line[1], float(line[7]), 0
            x = 0
            for symbol in sequence:
                try:
                    data[x][symbol].append(y)
                    
                except KeyError:
                    data[x][symbol] = [y]
                x += 1
        
        elif len(line) > 4 and 'multiple_map' in mode:
            
            if 'wild' in mode:
                sequence, y, empty = line[1], float(line[6]), 0
                x = 0
                
                for symbol in sequence:
                
                    if data[x].has_key(symbol):
                        data[x][symbol].append(y)
                    
                    else:
                        data[x][symbol] = [y]
                    x += 1
    
            elif line[0] != 'NA-NA':
                location = map(int, line[0].split('-')[0].split(','))
                id = line[0].split('-')[1].split(',')
                y, empty = float(line[6]), 0
                
                for x in location:
                    symbol = id.pop(0)
                    if data[x].has_key(symbol):
                        data[x][symbol].append(y)
                        
                    else:
                        data[x][symbol] = [y]
                    
        elif len(line) > 4 and 'difference' in mode and line[2] != 'NA':
            location = map(int, line[2].split(','))
            id = line[3].split(',')
            
            if 'difference_log2' in mode:
                y, empty = float(line[10]), 'NA'
                
            if 'difference_wt' in mode:
                y, empty = float(line[11]), 'NA'
                
            for x in location:
                symbol = id.pop(0)
                if data[x].has_key(symbol):
                    data[x][symbol].append(y)
                else:
                    data[x][symbol] = [y]
            
        elif len(line) > 4 and not line[4] == 'NA' and 'difference' not in mode:
            location = map(int, line[4].split(','))
            id = line[5].split(',')
            
            if mode == 'counts': 
                y, empty = float(line[7]), 0
            
            elif mode == 'unique':
                y, empty = 1, 0
                            
            elif mode == 'ratios': 
                y, empty = float(line[7]), "NA"
    
            elif mode == 'dnaratio':
                y, empty = float(line[14]), "NA"
    
            elif mode == 'dnastdev':
                y, empty = float(line[15]), "NA"
    
            elif mode == 'rosetta':
                y, empty = float(line[13]), "NA"
    
            for x in location:
                symbol = id.pop(0)
                try:
                    data[x][symbol].append(y)
                
                except KeyError:
                    data[x][symbol] = [y]
    
    abssum = 0
    maxabssum = 0
    
    if 'sum_absval' in mode:
        for x in range(0,size):
            
            for symbol in symbols:
                values = data[x][symbol]
                abssum = sum(map(abs, values))
                data[x][symbol] = abssum
                
                if maxabssum < abssum:
                    maxabssum = abssum
                
        for x in range(0,size):
            string = [str(x)]
            
            for symbol in symbols:
                
                if data[x][symbol] == 0:
                    string.append(empty)
                
                else:
                    string.append(data[x][symbol]/maxabssum)
            
            print >> f_outfile, '\t'.join(map(str, string))
                
    else:
        for x in range(0,size):
            string = [str(x)]
            for symbol in symbols:
                values = data[x][symbol]
                if values == []:
                    string.append(empty)
                    
                elif mode in ["counts", "unique", "wild"] or 'sum' in mode or 'wild_sum' in mode or 'counts' in mode: 
                    string.append(sum(values))
                
                elif mode in ["ratios", "rosetta"] or 'average' in mode or 'wild_average' in mode:
                    string.append(float(sum(values))/len(values))
                
                elif mode in ["dnaratio", "dnastdev"]:
                    string.append(sum(values))
                
            print >> f_outfile, '\t'.join(map(str, string))
    
    f_outfile.close()
    return(0)
            
if __name__ == '__main__':
    
    import optparse
    print time.asctime(time.localtime())    

    parser = optparse.OptionParser()
    parser.add_option('--molecule', action = 'store', type = 'string', dest = 'molecule', help = 'DNA or PRO')
    parser.add_option('--path', action = 'store', type = 'string', dest = 'path', help = 'path from script to files')
    parser.add_option('--infile', action = 'store', type = 'string', dest = 'infile', help = 'input file, is either mapCounts or mapRatios data')
    parser.add_option('--mode', action = 'store', type = 'string', dest = 'mode', help = 'should I use a per sequence counts column?')
    parser.add_option('--size', action = 'store', type = 'int', dest = 'size', help = 'length of input sequence (DNA or PRO)')
    parser.add_option('--local', action = 'store', type = 'string', dest = 'local', help = 'Is this a local (L) run or should an SGE (SGE) job be scheduled?')
    (option, args) = parser.parse_args()
    
    main(option.path, option.infile, option.molecule, option.mode, option.size, option.local)
    
    print time.asctime(time.localtime())
